grostl/src/grostl.rs

```diff
@@ -0,0 +1,492 @@
+            let shift = shifts[i] as usize;
+            if shift == 0 {
+                continue;
+            }
+            let d = gcd(shift, cols);
+            for j in 0..d {
+                let mut k = j;
+                let tmp = matrix[i][k];
+                loop {
+                    let pos = k.wrapping_add(shift) % cols;
+                    if pos == j {
+                        break
+                    }
+                    matrix[i][k] = matrix[i][pos];
+                    k = pos;
+                }
+                matrix[i][k] = tmp;
+            }
+        }
+    }
+
+    fn mix_bytes(
+        &self,
+        matrix: &Matrix<U8, Quot<BlockSize, U8>>,
+    ) -> Matrix<U8, Quot<BlockSize, U8>> {
+        matrix.mul_array(&B)
+    }
+
+    fn finalize(self) -> GenericArray<u8, OutputSize> {
+        let a = xor_generic_array(&self.p(&self.state), &self.state);
+        GenericArray::clone_from_slice(
+            &a[a.len() - OutputSize::to_usize()..],
+        )
+    }
+}
+
+impl<OutputSize, BlockSize> Default for Grostl<OutputSize, BlockSize>
+    where OutputSize: ArrayLength<u8>,
+          BlockSize: ArrayLength<u8> + Div<U8>,
+          BlockSize::ArrayType: Copy,
+          Quot<BlockSize, U8>: ArrayLength<u8>,
+{
+    fn default() -> Self { Self::new() }
+}
+
+impl<OutputSize, BlockSize> Digest for Grostl<OutputSize, BlockSize>
+    where OutputSize: ArrayLength<u8>,
+          BlockSize: ArrayLength<u8> + Div<U8>,
+          BlockSize::ArrayType: Copy,
+          Quot<BlockSize, U8>: ArrayLength<u8>,
+{
+    type OutputSize = OutputSize;
+    type BlockSize = BlockSize;
+
+    fn input(&mut self, input: &[u8]) {
+        let state = &mut self.state;
+        self.buffer.input(
+            input,
+            |b: &GenericArray<u8, BlockSize>| { state.compress(b); },
+        );
+    }
+
+    fn result(mut self) -> GenericArray<u8, Self::OutputSize> {
+        self.buffer.next(1)[0] = 128;
+        if self.buffer.remaining() >= 8 {
+            let block_bytes = self.block_bytes();
+            self.buffer.zero_until(block_bytes - 8);
+            {
+                let mut buf = self.buffer.next(8);
+                write_u64_be(&mut buf, (self.state.num_blocks + 1) as u64);
+            }
+            self.state.compress(self.buffer.full_buffer());
+        } else {
+            let block_bytes = self.block_bytes();
+            self.buffer.zero_until(block_bytes);
+            self.state.compress(self.buffer.full_buffer());
+
+            self.buffer.zero_until(block_bytes - 8);
+            {
+                let mut buf = self.buffer.next(8);
+                write_u64_be(&mut buf, (self.state.num_blocks + 1) as u64);
+            }
+            self.state.compress(self.buffer.full_buffer());
+        }
+
+        self.state.finalize()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{xor_generic_array, C_P, C_Q, Grostl, SHIFTS_P};
+    use generic_array::typenum::{U32, U64};
```

Comment:
On the second thought it will require to make tested methods public, they still will not be accessible to the crate user as `grostl` field will be private, but perhaps it's better to leave as is. Your choice.